include/boost/rational.hpp

```diff
@@ -148,33 +152,102 @@ class rational :
-    rational& assign(param_type n, param_type d);
+    template <class T>
+    typename enable_if_c<
+       rational_detail::is_compatible_integer<T, IntType>::value, rational &
+    >::type assign(const T& n, const T& d)
+    {
+       return *this = rational<IntType>(static_cast<IntType>(n), static_cast<IntType>(d));
+    }
+    //
+    // The following overloads should probably *not* be provided - 
+    // but are provided for backwards compatibity reasons only.
+    // These allow for construction/assignment from types that
+    // are wider than IntType only if there is an implicit
+    // conversion from T to IntType, they will throw a bad_rational
+    // if the conversion results in loss of precision or undefined behaviour.
+    //
+    template <class T>
+    rational(const T& n, typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value
+    >::type const* = 0)
+    {
+       assign(n, static_cast<T>(1));
+    }
+    template <class T>
+    rational(const T& n, const T& d, typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value
+    >::type const* = 0)
+    {
+       assign(n, d);
+    }
+    template <class T>
+    typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value,
+       rational &
+    >::type operator=(const T& n) { return assign(n, static_cast<T>(1)); }
+
+    template <class T>
+    typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value,
+       rational &
+    >::type assign(const T& n, const T& d)
+    {
+       if(!is_safe_narrowing_conversion(n) || !is_safe_narrowing_conversion(d))
+          BOOST_THROW_EXCEPTION(bad_rational());
```

Comment:
I don't see a better alternative. Disallowing construction from int altogether is too limiting as it would reject constants. If we go with an assert this makes the user do the work necessary to ensure that the assertion won't fire, at each assignment, instead of us doing the check once in rational. Throwing makes sense here.